docs/adr/adr-014-bootstrap-from-previous-peers.md

```diff
@@ -0,0 +1,470 @@
+                return shufflePeers(peers)
+        }
+        return ex, nil
+}
+ ```
+
+(_Code Snippet 2.b: Example of required changes to: go-header/p2p/exchange.go_)
+
+To pass the peer store to the `Exchange` (_and subsequently to the `peerTracker`_), we can rely on `WithPeerPersistence` option as follows:
+
+```diff
++   fx.Provide(newPeerStore),
+    fx.Provide(newHeaderService)
+...
+- func(cfg Config, network modp2p.Network) []p2p.Option[p2p.ClientParameters] {
++ func(cfg Config, network modp2p.Network, peerStore datastore.Datastore) []p2p.Option[p2p.ClientParameters] {
+        return []p2p.Option[p2p.ClientParameters]{
+            ...
+            p2p.WithChainID(network.String()),
++           p2p.WithPeerPersistence(peerStore),
+        }
+    },
+),
+     fx.Provide(newP2PExchange(*cfg)),
+```
+
+(_Code Snippet 3.b: Example of required changes to celestia-node/nodebuilder/header/module.go_)
+
+### Security Concerns: Long Range Attack
+
+With the ability to retrieve previously seen good peers, we should distinguish between those and the trusted peers for cases when we have expired local sync target (_i.e: subjective head_). On such cases, we would like to strictly sync and retrieve the subjective head from our trusted peers and no other, and for other syncing tasks, include the other stored good peers. To achieve this, we need to first include information about the expiry of the local sync target:
+
+```diff
+
+ // newP2PExchange constructs a new Exchange for headers.
+ func newP2PExchange(
+        lc fx.Lifecycle,
++        storeChainHead *header.ExtendedHeader,
+        bpeers modp2p.Bootstrappers,
+        network modp2p.Network,
+        host host.Host,
+        ...
+        exchange, err := p2p.NewExchange(...,
+                p2p.WithParams(cfg.Client),
+                p2p.WithNetworkID[p2p.ClientParameters](network.String()),
+                p2p.WithChainID(network.String()),
++               p2p.WithSubjectiveInitialization(storeChainHead.IsExpired),
+        )
+        if err != nil {
+                return nil, err
+```
+
+(_Code Snippet 3.c: Example of required changes to celestia-node/nodebuilder/header/constructors.go_)
+
+This piece information will allow the `Exchange` to distinguish between the two cases and use the correct peer set for each case:
+
+```diff
+ func (ex *Exchange[H]) Head(ctx context.Context) (H, error) {
+        log.Debug("requesting head")
++       // pool of peers to request from is determined by whether the head of the store is expired or not
++       if ex.Params.subjectiveInitialisation {
++              // if the local chain head is outside the unbonding period, only use TRUSTED PEERS for determining the head
++              peerset = ex.trustedPeers
++        } else {
++              // otherwise, node has a chain head that is NOT outdated so we can actually request random peers in addition to trusted
++              peerset = append(ex.trustedPeers, ex.params.peerstore.Get())
```

Comment:
Would it make sense on `Put` operation to store additional time.Now value and check the time passed since last store operation on `Load`, to identify if stored information is likely to be outdate and should be ignored?